parsing/lexer.mll

```diff
@@ -757,7 +780,16 @@ rule token = parse
   | ";"  { SEMI }
   | ";;" { SEMISEMI }
   | "<"  { LESS }
-  | "<[" { LESSLBRACKET }
+  | "<[" {
+      if !(syntax_mode.quotations) then
+        LESSLBRACKET
+      else
+        (* Put back the '[' and return just LESS *)
+        (lexbuf.lex_curr_pos <- lexbuf.lex_curr_pos - 1;
+         let curpos = lexbuf.lex_curr_p in
+         lexbuf.lex_curr_p <- { curpos with pos_cnum = curpos.pos_cnum - 1 };
```

Comment:
This backtracking does seem unavoidable. Since it also occurs in two other places, could we factor out this logic into a function.